CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc

```diff
@@ -533,18 +768,145 @@ void ttbbLepJetsAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetu
 	  b_GenLepton_eta = genttbarConeCat->begin()->lepton2().eta();
 	}
 
+	//---------------------------------------------------------------------------
+	// Generated particles: Lepton, neutrino and jets
+	//---------------------------------------------------------------------------
+	
+	edm::Handle<reco::GenParticleCollection> genParticles;
+	iEvent.getByToken(genToken_, genParticles);
+	
+	TLorentzVector Genlepton(0,0,0,0); 
+	TLorentzVector Gennu(0,0,0,0);
+
+	// Gen Status: http://home.thep.lu.se/~torbjorn/pythia81html/ParticleProperties.html
+	// abs(id) == 6  // t-Quark
+	// abs(id) == 5  // b-Quark
+	// abs(id) == 24 // W-Boson
+	// abs(id) == 15 // Tau
+	// abs(id) == 11 // Electron
+	// abs(id) == 13 // Muon
+
+	for (unsigned int i = 0; i < genParticles->size(); i++){
+
+	  const reco::Candidate & gp = (*genParticles)[i];
+	  const int id = gp.pdgId();
+
+	  // Only leptons
+	  if(abs(id) == 11 || abs(id) == 13 || abs(id) == 15 ||
+	     abs(id) == 16 || abs(id) == 14 || abs(id) == 12){
+
+	    const reco::Candidate *it     = 0;
+	    const reco::Candidate *mom    = 0;
+	    const reco::Candidate *itmom  = 0;
+	    const reco::Candidate *mommom = 0;
+
+	    int momid = id;
+	    it = (&gp);
+	    // This loop searches the particle's mother
+	    while(momid == id){
+	      if(it != 0){
+		mom = it->mother();
+		if(mom != 0) momid = mom->pdgId();
+		else momid = 0;
+		if(momid == id) it = mom;
+	      } // if(it != 0)
+	      else momid = 0;
+	    } // while(momid == id)
+
+	    int mommomid = momid;
+
+	    if(mom != 0){
+	      itmom = mom;
+	      // This loop searches the mother's mom of the particle
+	      while (mommomid == momid){
+		if(itmom !=0){
+		  mommom = itmom->mother();
+		  if(mommom != 0) mommomid = mommom->pdgId();
+		  else mommomid = 0;
+		  if(mommomid == momid) itmom = mommom->mother();
+		}
+		else mommomid = 0;
+	      } // if(mom != 0)
+	    } // while(mommomid == momid)
+
+	    if (abs(momid) == 24 && abs(mommomid) == 6){
+	      if (abs(id) == 13 || abs(id) == 11) Genlepton.SetPxPyPzE(gp.px(), gp.py(), gp.pz(), gp.energy());  
+
+	      if (abs(id) == 16 || abs(id) == 14 || abs(id) == 12) Gennu.SetPxPyPzE(gp.px(), gp.py(), gp.pz(), gp.energy());  
+	      
+	      if (abs(id) == 15){ // Taus 
+		for(unsigned int h = 0; h <  gp.numberOfDaughters(); h++) {
+		  const reco::Candidate *gd = gp.daughter(h);
+		  const int taudauid = gd->pdgId();
+		  if (abs(taudauid) == 13 || abs(taudauid) == 11) Genlepton.SetPxPyPzE(gd->px(), gd->py(), gd->pz(), gd->energy());  
+		} // for(taus' daughters)
+	      } // if(taus)
+	        
+	    } // if(t->W)
+	    
+	  }// if (mu || e || tau)
+	} // for(genParticles)
```

Comment:
This information is already implemented in GenTop collection. It is one of the time consuming area looping over all gen particles. We can drop it?